### ISSUE TITLE
Rule no-collection-size-mischeck failing with @typescript-eslint/parser

### ISSUE THREAD

**Issue by aarongoldenthal** [SonarSource/eslint-plugin-sonarjs#267]
**I want to report a bug.**

**Reproducer**

Given the following code:

```javascript
// index.js

// sonarjs/no-collection-size-mischeck
// works with default parser, not with with @typescript-eslint/parser
const foo = ['a', 'b'];
if (foo.length >= 0) {
    console.log('something');
}

const s = new Set();
if (s.size < 0) {
    console.log('something');
}

// sonarjs/no-ignored-return
// works with @typescript-eslint/parser, not with default parser
'hello'.lastIndexOf('e');
```

Running with the default eslint parser results in:

```sh
/app # npm run lint

> SonarjsTest@1.0.0 lint /app
> eslint *.js


/app/index.js
   6:5  error  Fix this expression; length of "foo" is always greater or equal to zero  sonarjs/no-collection-size-mischeck
  11:5  error  Fix this expression; size of "s" is always greater or equal to zero      sonarjs/no-collection-size-mischeck

✖ 2 problems (2 errors, 0 warnings)
```

Running with the same eslint configuration except adding `"parser": "@typescript-eslint/parser"` results in:

```sh
/app # npm run lint

> SonarjsTest@1.0.0 lint /app
> eslint *.js


/app/index.js
  17:1  error  The return value of "lastIndexOf" must be used sonarjs/no-ignored-return

✖ 1 problem (1 error, 0 warnings)
```

_The difference in `no-ignored-return` is expected given the latest documentation._

**Expected behavior**

Rule `no-collection-size-mischeck` should work with `@typescript-eslint/parser`.

**@typescript-eslint/parser:** 4.28.2
**eslint version:** 7.30.0
**eslint-plugin-sonarjs version:** 0.9.1
**typescript:** 4.3.5
**Node.js version:** v14.17.3 (Windows), v14.17.3 (Linux), v16.4.2 (Linux)

.eslintrc.js

```js
module.exports = {
    parserOptions: { ecmaVersion: 2019 },
    parser: '@typescript-eslint/parser',
    plugins: ['sonarjs'],
    rules: {
        'sonarjs/no-collection-size-mischeck': 'error',
        'sonarjs/no-ignored-return': 'error'
    }
};
```

**Comment by yassin-kammoun-sonarsource** [comment]
Hello @aarongoldenthal,

Because you are using `@typescript-eslint/parser`, the behaviour of `no-collection-size-mischeck` changes and uses type information to raise more accurate issues. To enable type information in your project, you need to set the [parserOptions.project](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsproject) option in `.eslintrc.js`.

**Comment by aarongoldenthal** [comment]
Hi @yassin-kammoun-sonarsource ,

I may not have explained the issue well.  I do not use the `@typescript-eslint/parser` today, I sue the default parser, and with the default parser it identifies the issues above for the `no-collection-size-mischeck` rule (taken from [the rule documentation](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-collection-size-mischeck.md)).

When I went to upgrade to v0.9.0, I had the issues identified in #265/#259/#262.  With those changes and v0.9.1, I tried using the `@typescript-eslint/parser` instead, as recommended, and it did resolve the issues in `no-ignored-return` noted above (also from [the rule documentatiom](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-ignored-return.md)). But, the examples from `no-collection-size-mischeck` are not flagged as errors with that perser.

**Comment by yassin-kammoun-sonarsource** [comment]
> [..] But, the examples from no-collection-size-mischeck are not flagged as errors with the @typescript-eslint/parser parser.

You didn't mention any issues with the examples from the rule documentation in your first message. This is somehow a different problem from what you initially reported to us. Therefore, I stand by what I said in my first answer with regard to the reproducer you shared.

Coming to the second problem you raised, the non-compliant examples - taken as they are - from the rule documentation are indeed not flagged when using `typescript-eslint/parser`. It's because those examples are missing variable declarations, e.g. `const someSet = new Set()`, for TypeScript compiler to infer types and for the rule to flag errors.

Long story short, there is no bug. Given your feedback, we just need to update the rule documentation.

**Comment by aarongoldenthal** [comment]
Thanks @yassin-kammoun-sonarsource, now I get it, and totally my bug. And this is still my favorite eslint plugin for performing meaningful checks  :)

For future reference for anyone else, `no-collection-size-mischeck` uses types if available, but if using the default parser where they are not then any statement evaluating any `.length`/`.size` property that is `>= 0` /`< 0` will be flagged, even if the statement is nonsense and those properties are undefined (since it doesn't know the types), e.g.

```js
const x = true;
if (x.length >= 0) { }

const y = 3;
if (y.length < 0) { }
```

will produce...

```sh
/app/index3.js
   1:5  error  Fix this expression; length of "x" is always greater or equal to zero  sonarjs/no-collection-size-mischeck
   4:5  error  Fix this expression; length of "y" is always greater or equal to zero  sonarjs/no-collection-size-mischeck
```

The real issue I had was Typescript setup for `typescript-eslint/parser`.  I had set up a tsconfig.json and `parserOptions.project` in my eslintrc as you recommended, but since it's a JavaScript project I wasn't actually running the Typescript compiler.  Once I ran `tsc` to generate type declarations and found it was not producing the correct types, I just had to update my tsconfig.json (e.g. setting `"lib": ["ES2019"]` so the `Set` type was identified).
